src/daft-local-execution/src/resource_manager.rs

```diff
@@ -0,0 +1,116 @@
+use std::sync::{Arc, Mutex};
+
+use common_error::{DaftError, DaftResult};
+use common_system_info::SystemInfoInternal;
+use lazy_static::lazy_static;
+use tokio::sync::Notify;
+
+lazy_static! {
+    pub static ref MEMORY_MANAGER: Arc<MemoryManager> = Arc::new(MemoryManager::new());
+}
+
+fn custom_memory_limit() -> Option<u64> {
+    let memory_limit_var_name = "DAFT_MEMORY_LIMIT";
+    if let Ok(val) = std::env::var(memory_limit_var_name) {
+        if let Ok(val) = val.parse::<u64>() {
+            return Some(val);
+        }
+    }
+    None
+}
+
+pub fn get_memory_manager() -> Arc<MemoryManager> {
+    MEMORY_MANAGER.clone()
+}
+
+pub struct MemoryPermit {
```

Comment:
You can name this `OwnedMemoryPermit`

and then in many cases you can also use:
```
MemoryPermit<'a> {
    bytes: u64,
    manager: &'a MemoryManager,
}
```
since if you are requesting the memory request in the tokio spawn, you don't need something with a static lifetime.